### PR TITLE
IPROTO-343 Allow null values for primitive types with PROTO3 syntax

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/IPROTO-343

### CODE REVIEW

File: integrationtests/src/test/java/org/infinispan/protostream/integrationtests/processor/marshaller/model/NullsAllowedModel.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+public class NullsAllowedModel {
+
+   public static final GeneratedSchema SCHEMA = new NullsAllowedSchemaImpl();
+
+   @ProtoField(1)
+   public String string;
+
+   @ProtoField(2)
+   public Integer integer;
```
Comment (pruivo) [inline integrationtests/src/test/java/org/infinispan/protostream/integrationtests/processor/marshaller/model/NullsAllowedModel.java:18]: If I change this to `int`, it fails with
```
Primitive field 'integer' of org.infinispan.protostream.integrationtests.processor.marshaller.model.NullsAllowedModel is not nullable so it should be either marked required or should have a default value, while processing org.infinispan.protostream.integrationtests.processor.marshaller.model.NullsAllowedModel.NullsAllowedSchema
```

File: core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java [hunk 2]
```diff
@@ -834,7 +834,7 @@ protected XClass getJavaTypeFromAnnotation(ProtoField annotation) {
     * type.
     */
    private Object getDefaultValue(XClass clazz, String fieldName, XClass fieldType, Type protobufType, String value, boolean isRepeated) {
-      if (protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2) {
+      if (protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2 || protoSchemaGenerator.allowNullFields()) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java:837]: this seems to do the trick:
```suggestion
      if (protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2 || (protoSchemaGenerator.allowNullFields() && !fieldType.isPrimitive())) {
```

**Review by pruivo — CHANGES_REQUESTED** [review]
Primitive types aren't working anymore. I suggested a possible fix :crossed_fingers:

**Comment by pruivo** [comment]
ps. would be nice to add a primitive type into the test too :+1:

**Comment by ryanemerson** [comment]
Good spot @pruivo. Fix and test updates pushed.

File: integrationtests/src/test/java/org/infinispan/protostream/integrationtests/processor/marshaller/model/NullsAllowedModel.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+   public static final GeneratedSchema SCHEMA = new NullsAllowedSchemaImpl();
+
+   @ProtoField(1)
+   public String string;
+
+   @ProtoField(2)
+   public Integer boxedInt;
+
+   @ProtoField(value = 3, defaultValue = "0")
```
Comment (pruivo) [inline integrationtests/src/test/java/org/infinispan/protostream/integrationtests/processor/marshaller/model/NullsAllowedModel.java:20]: is the `defaultValue="0"` a requirement? Or can we get rid of it? :smile:

Comment (ryanemerson) [reply integrationtests/src/test/java/org/infinispan/protostream/integrationtests/processor/marshaller/model/NullsAllowedModel.java]: Argh, that was a leftover from some local testing. Removed now.

**Comment by ryanemerson** [comment]
@pruivo Should be good now. I updated the testsuite and `@ProtoSchema` docs to reflect that the null behaviour also applies to Enums and `byte[]`.

**Review by pruivo — APPROVED** [review]
LGTM. Tested with Keycloak (https://github.com/pruivo/keycloak/commit/86a0614e3825a475eedea73e7be8374254daadf3)

**Comment by pruivo** [comment]
merged! thanks @ryanemerson !

**Pull request merged**